dhall-lsp-server/src/Dhall/LSP/Backend/Completion.hs

```diff
@@ -0,0 +1,186 @@
+buildCompletionContext' context values (Let (Binding x mA a :| []) e)
+  -- We prefer the actual value over the annotated type in order to get
+  -- 'dependent let' behaviour whenever possible.
+  | Right _A <- typeWithA absurd context a
+  , Right kind <- typeWithA absurd context _A =
+    let _A' = normalize _A
+        kind' = normalize kind
+
+        context' = fmap (shift 1 (V x 0)) $ insert x _A' context
+
+        a' = normalize a
+
+        e' | Const k <- kind', k `elem` [Kind, Sort] =
+             subst (V x 0) a' e  -- 'dependent let'
+           | otherwise = e
+
+        values' | Const k <- kind', k `elem` [Kind, Sort] =
+                  fmap (shift 1 (V x 0)) $ insert x a' values
+                | otherwise = fmap (shift 1 (V x 0)) $ insert x holeExpr values
+
+    in buildCompletionContext' context' values' e'
+
+  -- fall back to annotated type if body doesn't type check; can't do 'dependent let'
+  | Just _A <- mA
+  , Right _ <- typeWithA absurd context _A =
+    let _A' = normalize _A
+        context' = fmap (shift 1 (V x 0)) $ insert x _A' context
+        values' = fmap (shift 1 (V x 0)) $ insert x holeExpr values
+
+    in buildCompletionContext' context' values' e
+
+  -- if nothing works, only remember the name (but bind to `holeExpr`)
+  | otherwise =
+    let context' = fmap (shift 1 (V x 0)) $ insert x holeExpr context
+        values' = fmap (shift 1 (V x 0)) $ insert x holeExpr values
+
+    in buildCompletionContext' context' values' e
+
+buildCompletionContext' context values (Lam x _A b) =
+  let _A' | Right _ <- typeWithA absurd context _A = normalize _A
+          | otherwise = holeExpr
+
+      context' = fmap (shift 1 (V x 0)) $ insert x _A' context
+      values' = fmap (shift 1 (V x 0)) $ insert x holeExpr values
+
+    in buildCompletionContext' context' values' b
+
+buildCompletionContext' context values (Pi x _A b) =
+  let _A' | Right _ <- typeWithA absurd context _A = normalize _A
+          | otherwise = holeExpr
+
+      context' = fmap (shift 1 (V x 0)) $ insert x _A' context
+      values' = fmap (shift 1 (V x 0)) $ insert x holeExpr values
+
+    in buildCompletionContext' context' values' b
+
+-- catch-all
+buildCompletionContext' context values _ = CompletionContext context values
+
+-- Helper. Given `Dhall.Context.toList ctx` construct the corresponding variable
+-- names.
+contextToVariables :: [(Text, Expr Src X)] -> [Var]
+contextToVariables  [] = []
+contextToVariables ((name, _) : rest) =
+  V name 0 : map (inc name) (contextToVariables rest)
+ where inc x (V y i) | x == y = V x (i + 1)
+                     | otherwise = V y i
+
+-- | Complete identifiers from the given completion context.
+completeFromContext :: CompletionContext -> [Completion]
+completeFromContext (CompletionContext context _) =
+  let context' = toList context
+      completeReserved keyword
+        | Right expr <- exprFromText "" keyword
+        , Right typ <- typeOf (fmap undefined expr) =
```

Comment:
Instead of `fmap undefined expr`, you can do:

```haskell
typeOf (do _ <- expr; holeExpr)
```

This is because `Expr` is a `Monad` where `>>=` replaces all occurrences of `Embed`

Reply:
Neat!